src/esphome/sensor/mpr121_sensor.cpp

```diff
@@ -0,0 +1,90 @@
+#include "esphome/defines.h"
+#ifdef USE_MPR121
+
+#include "esphome/sensor/mpr121_sensor.h"
+#include "esphome/log.h"
+#include <bitset>
+
+ESPHOME_NAMESPACE_BEGIN
+
+namespace sensor {
+
+static const char *TAG = "sensor.mpr121";
+
+MPR121SensorChannel::MPR121SensorChannel(uint8_t channel, float value) {
+  this->channel = channel;
+  this->value = value;
+}
+
+MPR121Sensor::MPR121Sensor(const std::string &name) : Sensor(name) {
+  this->state = 0.0;
+ }
+
+void MPR121Sensor::process(uint16_t *data, uint16_t *last_data) {
+  //process depend on the sensor_type_
+  switch(this->sensor_type_) {
+    case MPR121_SENSOR_TYPE_SLIDER:
+      this->process_slider_(data);
+      break;
+    case MPR121_SENSOR_TYPE_ROTARY:
+      this->process_rotary_(data, last_data);
+      break;
+    default:
+      ESP_LOGE(TAG, "'%s': Unknown sensor type %d", this->name_.c_str(), this->sensor_type_);
+  }
+}
+
+void MPR121Sensor::process_slider_(uint16_t *data) {
+  uint8_t count = 0;
+  float value = 0;
+  for(auto *chan : this->channels_) {
+    if (*data & (1 << chan->channel)) {
+      value += chan->value;
+      count++;
+    }
+  }
```

Comment:
Ok so can you explain how this slider calculation works, how the slider is connected in comments? This feels a bit like black magic and I would like to understand it so that it can be maintained.

Reply:
ok done that. It realy simple.
In this method we check what channels where touched, then publish the average value for all toched channels.
The Rotary sensor (or Wheel sensor, I will rename this) is much more complex.